### PR TITLE
Add hyperparameter tutorial

### PR DESCRIPTION
**Pull request by kuimoani**
This is a tutorial that applies hyperparameters to existing MLOps pipeline code.
It is associated with the [previous sample](https://github.com/liupeirong/MLOpsManufacturing/tree/main/samples/wrapping-existing-ml-scripts-tutorial).

### CODE REVIEW

File: samples/hyperparameter-tutorial/.gitattributes [hunk 1]
```diff
@@ -0,0 +1,2 @@
+samples/yesno/src/**    eol=lf
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/.gitattributes]: Doesn't this need to be changed to hyperparameter-tutorial?

File: samples/hyperparameter-tutorial/src/steps/01_prep_data.py [hunk 2]
```diff
@@ -0,0 +1,37 @@
+from azureml.core.run import Run, _OfflineRun
+from azureml.core import Workspace
+from sklearn.datasets import load_diabetes
+import numpy as np
+import argparse
+import os
+
+# Get context
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/src/steps/01_prep_data.py:8]: Looks like run context and workspace are not used in this file?! Can we remove them?

File: samples/hyperparameter-tutorial/src/steps/02_train.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+from azureml.core import Workspace
+from sklearn.linear_model import Ridge
+from sklearn.metrics import mean_squared_error, r2_score
+from sklearn.model_selection import train_test_split
+import os
+import joblib
+import numpy as np
+import argparse
+
+# Get context
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/src/steps/02_train.py:11]: same comment as for 01_prep_data

File: samples/hyperparameter-tutorial/README.md [hunk 4]
```diff
@@ -0,0 +1,169 @@
+# Overview
+
+This sample guides you on how to create a pipeline of Hyper Parameter tune with  Azure Machine Learning.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
This sample guides you on how to create a Azure Machine Learning pipeline with Hyper Parameter tuning.
```

File: samples/hyperparameter-tutorial/README.md [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+
+__What does this sample demonstrate__:
+
+* Create Hyper parameter tune step into the Azure ML Pipeline
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
* How to use the Hyper Parameter tune step in the Azure ML Pipelines
```

File: samples/hyperparameter-tutorial/README.md [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+* Demo how to handle the metrics
+
+__What doesn't this sample demonstrate__:
+
+* Basic about Azure ML Pipeline
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
* Basics about Azure ML Pipeline
```

File: samples/hyperparameter-tutorial/README.md [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+
+## Introduction
+
+> Wrapping existing ML scripts Tutorial: https://github.com/liupeirong/MLOpsManufacturing/tree/main/samples/wrapping-existing-ml-scripts-tutorial
+
+In the example above, we completed the basic MLOps pipeline. This time, we're going to modify the existing code to change it to a pipeline that uses hyperparameter tuning and get the best model.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
In the example above, we created a basic Azure ML Pipeline. This time, we're going to modify the existing code to change it to a pipeline that uses hyperparameter tuning to identify the best model.
```

File: samples/hyperparameter-tutorial/README.md [hunk 8]
```diff
@@ -0,0 +1,169 @@
...
+
+The above example was [Linear Regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html), but for the hyperparameter tuning example, we will modify it to [Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) and use _alpha(Regularization strength)_ as a hyperparameter.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
The previous example was about [Linear Regression](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LinearRegression.html), for the hyperparameter tuning example, we will modify it to use [Ridge](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.Ridge.html) and _alpha(Regularization strength)_ as a hyperparameter.
```

File: samples/hyperparameter-tutorial/README.md [hunk 9]
```diff
@@ -0,0 +1,169 @@
...
+
+Questions:
+
+- How to change an existing MLOps pipeline to use Hyper Parameters?
+- How to get the best model from the training using Hyper Parameters?
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
- How to retrieve the best model from the pipeline run using Hyper Parameters?
```

File: samples/hyperparameter-tutorial/README.md [hunk 10]
```diff
@@ -0,0 +1,169 @@
...
+
+We will guide you step by step through both answers to these questions. 
+
+# Getting Started
+
+## Understanding Concepts
+
+Hyperparameter tuning is one of the most challenging issues in Machine Learning, but using the Azure Cloud can easily solve this problem.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
Hyperparameter tuning is one of the most challenging issues in Machine Learning, but using the Azure Machine Learning makes this problem a lot easier to solve.
```

File: samples/hyperparameter-tutorial/README.md [hunk 11]
```diff
@@ -0,0 +1,169 @@
...
+
+Parameter sampling can specify hyperparameter values in a range or random, adjust the number of compute instances to train at the same time, or lower performance training can be early terminated, saving costs.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
Parameter sampling can specify hyperparameter values in a range or random, adjust the number of compute instances to train at the same time, or training runs with low performance can be early terminated, saving costs.
```

File: samples/hyperparameter-tutorial/README.md [hunk 12]
```diff
@@ -0,0 +1,169 @@
...
+
+If you have an existing Azure ML pipeline code, you just need to work on the below.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
If you have an existing Azure ML pipeline code, you just need to do the following:
```

File: samples/hyperparameter-tutorial/README.md [hunk 13]
```diff
@@ -0,0 +1,169 @@
...
+
+  - Set hyperparameter values and policies
+  - Add the hyperparameters to be passed over
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
  - Add the hyperparameters that need to be passed
```

File: samples/hyperparameter-tutorial/README.md [hunk 14]
```diff
@@ -0,0 +1,169 @@
...
+  - Modify the existing train step to hyperdrive step
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
  - Convert the existing train step to a hyperdrive step
```

File: samples/hyperparameter-tutorial/README.md [hunk 15]
```diff
@@ -0,0 +1,169 @@
...
+...
+# train
+model = Ridge(alpha=args.alpha)
+model.fit(X_train, y_train)
+...
+```
+
+## Modifying the pipeline build
+
+Now in the code that creates a pipeline, we change the existing training step to the hyperdrive step.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
Now in the code that creates the pipeline, we change the existing training step to the hyperdrive step.
```

File: samples/hyperparameter-tutorial/README.md [hunk 16]
```diff
@@ -0,0 +1,169 @@
...
+
+## Running the pipeline
+
+When the pipeline runs, it trains according to the number of instances in the _Compute Cluster_ and MAX_TOTAL_RUNS, MAX_CONCURRENT_RUNS in `hyperparams.py`.
+
+$ python -m ml_service.pipelines.run_pipeline
+
+> Using Compute Cluster may take a while to provision at the first time.
+
+When it runs normally, you can see the screenshot below.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
When it runs normally, the result should look similar like you can see the screenshot below.
```

File: samples/hyperparameter-tutorial/README.md [hunk 17]
```diff
@@ -0,0 +1,169 @@
...
+
+![image-20210310163438863](docs/image-20210310163438863.png)
+
+In the Hyper drive step, you can see the trend of changes in hyperparameter values and primary metric values in the graph. Through this, you can analyze how hyperparameters affect each model.
+
+![image-20210310164054229](docs/image-20210310164054229.png)
+
+After finishing the trains, the model with max of _r2_ values as set in the primary metric in `ml_service/pipelines/hyperparams.py` is registered as the best model.
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
After finishing the trainings, the model with max of _r2_ values as set in the primary metric in `ml_service/pipelines/hyperparams.py` is registered as the best model.
```

File: samples/hyperparameter-tutorial/README.md [hunk 18]
```diff
@@ -0,0 +1,169 @@
...
+
+# Conclusion
+
+We have looked at the basics of changing the existing MLOps code to a pipeline for hyperparameter tuning. From now on, start working on increasing the accuracy of your model by the hyperparameter tune you want!
```
Comment (h2floh) [inline samples/hyperparameter-tutorial/README.md]:
```suggestion
We have looked at the basics of changing the existing Azure ML code into a pipeline for hyperparameter tuning. From now on, start working on increasing the accuracy of your model by the hyperparameter tune you want!
```

**Review by h2floh — COMMENTED** [review]
Hey @kuimoani , thanks for contributing a HyperParameter tutorial. Please check the comments and add you tutorial to the new sample [overview page](https://github.com/liupeirong/MLOpsManufacturing/tree/main/samples).

File: samples/hyperparameter-tutorial/.gitattributes (hunk 1 above)
Comment (kuimoani) [reply samples/hyperparameter-tutorial/.gitattributes]: fixed!

File: samples/hyperparameter-tutorial/src/steps/01_prep_data.py (hunk 2 above)
Comment (kuimoani) [reply samples/hyperparameter-tutorial/src/steps/01_prep_data.py:8]: No, these are needed.
There are used to handle arguments and directory.

File: samples/hyperparameter-tutorial/src/steps/02_train.py (hunk 3 above)
Comment (kuimoani) [reply samples/hyperparameter-tutorial/src/steps/02_train.py:11]: It is needed as well.

File: samples/hyperparameter-tutorial/src/steps/01_prep_data.py (hunk 2 above)
Comment (h2floh) [reply samples/hyperparameter-tutorial/src/steps/01_prep_data.py:8]: Still can't find any references in the prep_data script to `run.` or `ws.`. Doesn't argparse handle the arguments?

File: samples/hyperparameter-tutorial/src/steps/02_train.py (hunk 3 above)
Comment (h2floh) [reply samples/hyperparameter-tutorial/src/steps/02_train.py:11]: Ok, found the reference to `run.` :)

**Comment by h2floh** [comment]
@kuimoani , please also add your sample to the [sample overview tables](https://github.com/liupeirong/MLOpsManufacturing/tree/main/samples), if you are unsure about what to enter, either leave it blank or 'TBD' (To Be Defined), that we can update it later on. Thanks

File: samples/hyperparameter-tutorial/src/steps/01_prep_data.py (hunk 2 above)
Comment (kuimoani) [reply samples/hyperparameter-tutorial/src/steps/01_prep_data.py:8]: Oh! I'm sorry. I misunderstood it. As you said, _run_ and _ws_ are not used.
However, I think it would be better to be left here because _run_ and _ws_ could be a kind of pattern to re-use context within each script of steps.

File: samples/README.md [hunk 19]
```diff
@@ -42,6 +44,7 @@ principles which are covered
 |Sample|[Cost Optimization](https://docs.microsoft.com/en-us/azure/architecture/framework/cost/overview)|[Operational Excellence](https://docs.microsoft.com/en-us/azure/architecture/framework/devops/overview)|[Performance Efficiency](https://docs.microsoft.com/en-us/azure/architecture/framework/scalability/overview)|[Reliability](https://docs.microsoft.com/en-us/azure/architecture/framework/resiliency/overview)|[Security](https://docs.microsoft.com/en-us/azure/architecture/framework/security/security-principles)|
 |----|----|----|----|----|----|
 |[Appendable AML Pipeline Step Template](./appendable-template)|Pay for consumption|||||
+|[Hyperparameter Tutorial](./hyperparameter-tutorial)|Pay for consumption|Optimized the model|Monitor and optimize (planned)|||
```
Comment (h2floh) [inline samples/README.md]:
```suggestion
|[Hyperparameter Tutorial](./hyperparameter-tutorial)|Pay for consumption|Optimize build and release processes||||
```

**Review by h2floh — APPROVED** [review]
Thanks @kuimoani ! LGTM
